### ISSUE TITLE
Reliable Importing of assets is not possible

### ISSUE THREAD

**Issue by exophunk** [verbb/feed-me#451]
### Description

I am struggling with a reliable asset import for weeks now and with every update, it seems it stops to work for another reason:

#### Importing into Asset Element type directly:
1. Providing a filename only like `image1.jpg` does not work at all:
  - I place a file `image1.jpg` into the corresponding volume folder
  - I import from a json structure like this `{ filename: "1.jpg" }`
  - Nothing happens. FeedMe silently fails and no errors are thrown. No assets are created.
2. Providing a absolute url and selecting `Create asset from URL` works, but:
  - This is ugly, I have the file on the internal file system and don't want to go via a external, public URL
  - The images get re-uploaded every single time, evern if the files do not change at all. Importing 1900 images take about 30 minutes, because it does not update the delta only. Is it possible to somehow use `compareContent: true` on assets as well?
3. After months, there is still absolutly no documentation on how to import assets at all. This is okay for a open source project, but for a paid – and quiet expensive – feature, I somewhat expect it to work.

#### Connecting entries to assets
1. When importing entries with an asset field, I import a filename only like `image1.jpg` into the asset field of the entry.
2. Nothing happens. (asset with this filename exists) No relations are created, No asset is added. FeedMe silently fails with no errors. This seem to have worked on previous versions!
3. Again, providing an absolute URL and `Create asset from URL` works and assets are created on the fly. But that is not what i want to do! I have existing, already imported assets that I want to connect.
4. Using anything else than the filename does not work as well. I didn't found any way to make importing assets work.

### Steps to reproduce

1. Import assets or entries with asset fields
2. Try to make it work with non-absolute URLs
3. Good luck

### Additional info

- Plugin version: 3.1.0
- Craft version: 3.1.3

**Comment by exophunk** [comment]
I think I found one issue, why the "Connecting entries to assets" did not work in my case:

When [parsing the asset field](https://github.com/verbb/feed-me/blob/craft-3/src/fields/Assets.php#L40), it tries to get the folder by its UID. In my case, the Volume UID was passed and it tried to match the `volumefolder` by this UID. In my case, the folder had a different UID than the volume and so, no folder was found.

https://github.com/verbb/feed-me/blob/craft-3/src/fields/Assets.php#L58

I don't know if this UID's should be the same normally and this was just bugged in my system due to updating to Craft 3.1.

**Comment by rogerdawkins** [comment]
I'm having the same issue. I have a Feed Me import setup that was working fine a few weeks ago. In my case it is a CSV file and I have an asset field containing "123.jpg". The asset is in the relevant asset folder and the import runs without any errors and all other fields are imported but the Asset never gets assigned to the matching filename in the Assets folder. I presume I could have the same issue with UID's not matching but not sure how to check that without debugging Feed Me.

I'm going to try the full url and let Feed Me create the asset, but there is definitely an issue.

**Comment by engram-design** [comment]
Thanks for the detailed post @exophunk and sorry to hear your issues have been so ongoing. We were reasonably confident things were working much better as of the latest release due to a good refactor, but evidently not.

You've hit the nail on the head - it was indeed the UID lookup - a change made in Craft 3.1.  There have been numerous changes in Craft 3.1, which is indeed frustrating. This should fix matching against existing assets by providing the filename (this is for asset fields). In fact, I think this was part of a PR we just didn't test properly...

I have written up documentation on asset element importing. I hope that helps to clarify some of the items - you're right, we've been lazy at getting this sorted.

When importing assets providing just the filename, it certainly will update an existing asset by that name. Use case - say you already have all your assets in Craft, but you've added some custom fields, and want to update those. You provide the filename, and those custom fields' content, and away it goes to update. However, it won't upload it for you from a local path. 

This has been changed through some refactoring on the field mapping screen. You now have a "Path or URL" field when you set your feed to "Create new elements". This makes a lot more sense than the previous "Create from URL", which is a carry-over from the Asset Field implementation. It also addresses another issues where people wanted to modify the filename. A migration should sort this out for you, but best to double check your feed in any case.

Previously, I recommended a much, much more performant to copy the image files into your volume folder, and then run the "Reindex assets" utility, which will create them as assets in Craft. Basically, this scans the volume directory and adds/removes assets based on the files in there. I've mentioned this in the docs. But - this isn't always the desired workflow (say some other service uploads images into your volume, and feed me needs to add those to Craft.

I'm not sure how we can use `compareContent: true` with an asset, at least reliably. Sure we can compare size, but we'd just be opening it up to "assets aren't updating". Its something we'll consider for sure though. There is a caching mechanism in place that images are downloaded into the temporary cache directory, so they're not continually downloaded.

These changes have been made in the craft-3 branch and will be in the next release.
